### PR TITLE
#742 - Workspace Scan API

### PR DESCRIPTION
**Pull request by gisjedi**
Added ability to construct Workspace Scan models and kick off dry run and full ingest process from them via REST API. Documentation is also added for Scan to the REST and Architecture sections.

### CODE REVIEW

File: scale/ingest/models.py [hunk 1]
```diff
@@ -723,7 +721,44 @@ def get_details(self, scan_id):
...
+        atomic transaction.
+
+        :param scan_id: The unique identifier of the Scan process.
+        :type scan_id: int
+        :param dry_run: Whether the scan will execute as a dry run
+        :type dry_run: bool
+        :returns: The new Scan process
+        :rtype: :class:`ingest.models.Scan`
+        """
+
```
Comment (dfaller) [inline scale/ingest/models.py]: We need to think about overwriting an old dry or normal job. We probably don't want to re-run a normal job, not sure about dry runs.

File: scale/ingest/urls.py [hunk 2]
```diff
@@ -12,6 +12,13 @@
     url(r'^ingests/(?P<ingest_id>\d+)/$', views.IngestDetailsView.as_view(), name='ingest_details_view'),
     url(r'^ingests/(?P<file_name>[\w.]{0,250})/$', views.IngestDetailsView.as_view(), name='ingest_details_view'),
 
+    # Scan views
+    url(r'^scans/$', views.ScansView.as_view(), name='scans_view'),
+    url(r'^scans/$', views.ScansView.as_view(), name='scans_view'),
```
Comment (dfaller) [inline scale/ingest/urls.py]: Duplicate line

File: scale/ingest/views.py [hunk 3]
```diff
@@ -42,10 +44,16 @@ def list(self, request):
...
         file_name = rest_util.parse_string(request, 'file_name', required=False)
         order = rest_util.parse_string_list(request, 'order', required=False)
 
-        ingests = Ingest.objects.get_ingests(started, ended, ingest_statuses, strike_ids, file_name, order)
+        ingests = Ingest.objects.get_ingests(started=started, ended=ended,
+                                             statuses=ingest_statuses,
+                                             scan_ids=scan_ids,
+                                             strike_ids=strike_ids,
+                                             file_name=file_name,
+                                             order=order)
```
Comment (dfaller) [inline scale/ingest/views.py]: Need to update ingest REST API docs with the new parameter.

**Comment by gisjedi** [comment]
During dry runs and prior to an actual ingest, configuration changes should be allowed. Once you have launched the ingest job, no further dry runs or configuration changes will be allowed. We will continue to allow multiple dry runs and only the newest dry run job will be related to the Scan model.

File: scale/docs/rest/scan.rst [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+| name               | String            | Required | The identifying name of the Scan process used for queries.          |
++--------------------+-------------------+----------+---------------------------------------------------------------------+
+| title              | String            | Optional | The human readable display name of the Scan process.                |
++--------------------+-------------------+----------+---------------------------------------------------------------------+
+| description        | String            | Optional | A longer description of the Scan process.                           |
++--------------------+-------------------+----------+---------------------------------------------------------------------+
+| configuration      | JSON Object       | Required | JSON defining the Scan configuration.                               |
+|                    |                   |          | (See :ref:`architecture_scan_spec`)                                 |
++--------------------+-------------------+----------+---------------------------------------------------------------------+
```
Comment (gisjedi) [inline scale/docs/rest/scan.rst]: Nope, as that will be identified in the `ingest` parameter given to the `scan/{id}/process` API call.

File: scale/ingest/models.py [hunk 5]
```diff
@@ -626,15 +632,6 @@ def create_scan(self, name, title, description, configuration, dry_run=True):
...
-        scan_type = self.get_scan_job_type()
-        job_data = JobData()
-        job_data.add_property_input('Scan ID', unicode(scan.id))
-        job_data.add_property_input('Dry Run', str(dry_run))
-        event_description = {'scan_id': scan.id}
-        event = TriggerEvent.objects.create_trigger_event('SCAN_CREATED', None, event_description, now())
-        scan.job = Queue.objects.queue_new_job(scan_type, job_data, event)
-        scan.save()
-
         return scan
```
Comment (gisjedi) [inline scale/ingest/models.py]: Updated in docs.

File: scale/ingest/models.py (hunk 1 above)
Comment (gisjedi) [inline scale/ingest/models.py]: Dry runs are allowed to be overwritten, normal scans that perform ingests cannot be re-run.

File: scale/ingest/views.py (hunk 3 above)
Comment (gisjedi) [inline scale/ingest/views.py]: Done

File: scale/ingest/models.py [hunk 6]
```diff
@@ -656,6 +654,9 @@ def edit_scan(self, scan_id, title=None, description=None, configuration=None):
         """
 
         scan = Scan.objects.get(pk=scan_id)
```
Comment (dfaller) [inline scale/ingest/models.py]: We need to do a model lock here.

File: scale/ingest/models.py [hunk 7]
```diff
@@ -723,7 +726,47 @@ def get_details(self, scan_id):
...
+
+        :param scan_id: The unique identifier of the Scan process.
+        :type scan_id: int
+        :param dry_run: Whether the scan will execute as a dry run
+        :type dry_run: bool
+        :returns: The new Scan process
+        :rtype: :class:`ingest.models.Scan`
+        """
+
+        scan = Scan.objects.get(pk=scan_id)
```
Comment (dfaller) [inline scale/ingest/models.py]: Need a model lock here too.

**Pull request merged**